src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -350,61 +477,56 @@ protected virtual void ValidateCHash(OpenIdConnectProtocolValidationContext vali
             }
 
             var chash = cHashClaim as string;
             if (chash == null)
             {
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10326, validationContext.IdToken), typeof(OpenIdConnectProtocolInvalidCHashException), EventLevel.Error);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10306, idToken), typeof(OpenIdConnectProtocolInvalidCHashException), EventLevel.Error);
             }
 
             try
             {
-                ValidateHash(chash, validationContext.ProtocolMessage.Code, validationContext.IdToken.Header.Alg);
+                ValidateHash(chash, validationContext.ProtocolMessage.Code, idToken.Header.Alg);
             }
             catch(OpenIdConnectProtocolException ex)
             {
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10329, validationContext.IdToken), typeof(OpenIdConnectProtocolInvalidCHashException), EventLevel.Error, ex);
+                LogHelper.Throw(ex.Message, typeof(OpenIdConnectProtocolInvalidCHashException), EventLevel.Error, ex);
             }
         }
 
         /// <summary>
         /// Validates the 'token' according to http://openid.net/specs/openid-connect-core-1_0.html
         /// </summary>
+        /// <param name="idToken">The <see cref="JwtSecurityToken"/> to validate.</param>
         /// <param name="validationContext">a <see cref="OpenIdConnectProtocolValidationContext"/> that contains the protocol message to validate.</param>
+        /// <exception cref="ArgumentNullException">if 'idToken' is null.</exception>
         /// <exception cref="ArgumentNullException">if 'validationContext' is null.</exception>
         /// <exception cref="OpenIdConnectProtocolInvalidAtHashException">if the validationContext contains a 'token' and there is no 'at_hash' claim in the id_token.</exception>
         /// <exception cref="OpenIdConnectProtocolInvalidAtHashException">if the validationContext contains a 'token' and the 'at_hash' claim is not a string in the 'id_token'.</exception> 
-        /// <exception cref="OpenIdConnectProtocolInvalidAtHashException">if the <see cref="OpenIdConnectProtocolValidationContext.IdToken"/> is null and <see cref="OpenIdConnectProtocolValidationContext.ProtocolMessage.IdToken"/> != null. This indicates the id_token was not validated.</exception> 
-        protected virtual void ValidateAtHash(OpenIdConnectProtocolValidationContext validationContext)
+        /// <exception cref="OpenIdConnectProtocolInvalidCHashException">if the 'at_hash' claim in the 'idToken' does not correspond to the 'token' in the <see cref="OpenIdConnectMessage"/> response.</exception> 
+        protected virtual void ValidateAtHash(JwtSecurityToken idToken, OpenIdConnectProtocolValidationContext validationContext)
```

Comment:
Not sure we should have the 'idToken' parameter. This could cause confusion as OpenIdConnectProtocolValidationContext.ValidatedIdToken could also be set, and it must be set in the Validatexyz methods.